File: src/core/markdown/opbuffer.ts
// Op buffer — the renderer-independent document representation from
// docs/decisions/007-performance-path.md section 4.
//
// Since P2.3 the buffer is *produced in Rust* (`src-tauri/markdown`) and
// arrives ready to replay: over IPC on the desktop, from the WebAssembly build
// of the same crate in the browser preview and the tests. This module only
// reads it. The encoder that used to live here scanned marked's HTML output and
// disappeared with marked (docs/decisions/008-parser-reference.md).
//
// OFFSET DECISION (the most common source of errors with this approach; see
// 007 section 4, "Fallstricke"):
//
//   `strings` stays a UTF-8 blob because that is compact and transferable.
//   Every offset and length stored in a TEXT op, an attribute entry or a
//   heading entry is a **UTF-16 code unit offset into the JS string that
//   `decodeStrings()` produces once per document**, not a byte offset.
//
// The Rust encoder carries the UTF-16 length forward while it appends pieces,
// which costs one addition per piece. The replay can then use
// `String.prototype.substring` without a byte-to-UTF-16 translation table.
// Byte offsets would have forced that table into the hot path — exactly the
// work this format exists to remove.
//
// Layout (all arrays are Uint32Array except `strings`):
//
//   ops       4 words per operation  [kind, a, b, c]
//               kind 0 OPEN   a = tagId      b = attrStart  c = attrCount
//               kind 1 CLOSE  a, b, c = 0
//               kind 2 TEXT   a = strOffset  b = strLen     c = 0
//   attrs     3 words per attribute  [nameId, strOffset, strLen]
//   strings   UTF-8 blob: attribute values, heading ids and texts, and the
//             HTML of fallback sections — never document text
//   text      UTF-8 blob: the document's text in document order, with a
//             separator between blocks. TEXT operations index into it, and the
//             search runs on it directly (007, P2.4)
//   sections  9 words per section    [opStart, opCount, flags,
//                                     hashLow, hashHigh, htmlOffset, htmlLen,
//                                     textStart, textLen]
//   headings  6 words per heading    [level, idOffset, idLen, sectionIndex,
//                                     textOffset, textLen]
//
// Deviations from 007, each with a reason:
//   * `sections` carries `flags`. A section that could not be encoded must be
//     distinguishable from a section that legitimately produced no operations,
//     and `[opStart, opCount]` cannot express that.
//   * `sections` carries a 64-bit content hash. The viewport recognizes
//     unchanged sections by it; the HTML string it used to compare disappeared
//     with the encoder (docs/decisions/008, section 5).
//   * `sections` carries the HTML of fallback sections, as a slice of the same
//     string blob. Raw HTML has no opcode and stays on the DOMPurify path.
//   * `headings` carries the heading text, which the outline needs and which no
//     longer travels as JSON beside the buffer.
//   * The text is a blob of its own. It is what the search reads, so it must be
//     contiguous and free of attribute values; and `sections` carries each
//     section's range in it, so a match maps back to a section in one lookup.
//
// Every OPEN has a matching CLOSE, void elements included. That keeps the
// replay a single stack without a void-element table of its own.

// Exactly the tag allowlist the sanitizer enforces; policy.ts imports it so the
// two can never drift. A tag outside this table has no id and is therefore not
// expressible in the buffer at all, rather than being removed afterwards.
// `src-tauri/markdown/src/opbuffer.rs` carries the same two tables; the round
// trip over the real parser in tests/opbuffer.test.ts is what keeps them aligned.
export const ALLOWED_TAGS = [
  'p',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'em',
  'strong',
  'del',
  's',
  'code',
  'pre',
  'blockquote',
  'ul',
  'ol',
  'li',
  'hr',
  'br',
  'a',
  'img',
  'table',
  'thead',
  'tbody',
  'tfoot',
  'tr',
  'th',
  'td',
  'input',
  'sup',
  'sub',
  'kbd',
  'details',
  'summary',
  'div',
  'span',
  'dl',
  'dt',
  'dd',
] as const;

// Exactly the attribute allowlist the sanitizer enforces. Attribute *values*
// stay subject to inspection during replay; see policy.ts.
export const ALLOWED_ATTR = [
  'id',
  'href',
  'src',
  'alt',
  'title',
  'class',
  'start',
  'type',
  'checked',
  'disabled',
  'colspan',
  'rowspan',
  'align',
  'width',
  'height',
  'open',
] as const;

export const OP_OPEN = 0;
export const OP_CLOSE = 1;
export const OP_TEXT = 2;
export const OP_WORDS = 4;
export const ATTR_WORDS = 3;
export const SECTION_WORDS = 9;
export const HEADING_WORDS = 6;
/** The section could not be encoded; the renderer must sanitize its HTML. */
export const SECTION_FALLBACK = 1;

export interface OpBuffer {
  ops: Uint32Array;
  attrs: Uint32Array;
  strings: Uint8Array;
  text: Uint8Array;
  sections: Uint32Array;
  headings: Uint32Array;
}

/** Both blobs, decoded once per document. */
export interface OpStrings {
  /** Attribute values, heading ids and texts, fallback HTML. */
  readonly strings: string;
  /** The document's text in document order. */
  readonly text: string;
}

// The five arrays are views on one received packet; nothing copies or
// transfers them any more, because the parser no longer runs in a worker.

/** Decodes both blobs. Call once per document, not once per section. */
export function decodeStrings(buffer: OpBuffer): OpStrings {
  const decoder = new TextDecoder();
  return {
    strings: decoder.decode(buffer.strings),
    text: decoder.decode(buffer.text),
  };
}

export function sectionCount(buffer: OpBuffer): number {
  return buffer.sections.length / SECTION_WORDS;
}

/** False when the section must be sanitized through the HTML path instead. */
export function sectionEncoded(buffer: OpBuffer, index: number): boolean {
  const base = index * SECTION_WORDS;
  return (
    base < buffer.sections.length &&
    (buffer.sections[base + 2] & SECTION_FALLBACK) === 0
  );
}

/**
 * Content identity of a section, for recognizing unchanged sections across
 * reloads. A 64-bit hash from the parser, not a comparison: holding the
 * previous document's buffers for an exact comparison would cost tens of
 * megabytes at 10 MiB (docs/decisions/008, section 5).
 */
export function sectionKey(buffer: OpBuffer, index: number): string {
  const base = index * SECTION_WORDS;
  return `${buffer.sections[base + 3]}:${buffer.sections[base + 4]}`;
}

/** The HTML of a fallback section; empty for encoded sections. */
export function sectionHtml(
  buffer: OpBuffer,
  strings: OpStrings,
  index: number,
): string {
  const base = index * SECTION_WORDS;
  return strings.strings.substring(
    buffer.sections[base + 5],
    buffer.sections[base + 5] + buffer.sections[base + 6],
  );
}

/** Where the section's text lies in the document text: [start, end). */
export function sectionTextRange(
  buffer: OpBuffer,
  index: number,
): [number, number] {
  const base = index * SECTION_WORDS;
  return [
    buffer.sections[base + 7],
    buffer.sections[base + 7] + buffer.sections[base + 8],
  ];
}

/**
 * Document-text offset of the section's first text run. Offsets inside a
 * section are kept relative to it, so a section reused across reloads keeps a
 * valid text-node map even though the document around it moved.
 */
export function sectionFirstText(buffer: OpBuffer, index: number): number {
  const base = index * SECTION_WORDS;
  const start = buffer.sections[base];
  const end = start + buffer.sections[base + 1];
  for (let i = start; i < end; i++)
    if (buffer.ops[i * OP_WORDS] === OP_TEXT)
      return buffer.ops[i * OP_WORDS + 1];
  return 0;
}

/** The section a document-text offset falls into, or -1. */
export function sectionAt(buffer: OpBuffer, offset: number): number {
  let low = 0;
  let high = sectionCount(buffer) - 1;
  while (low <= high) {
    const mid = (low + high) >> 1;
    const base = mid * SECTION_WORDS;
    if (offset < buffer.sections[base + 7]) high = mid - 1;
    else if (offset >= buffer.sections[base + 7] + buffer.sections[base + 8])
      low = mid + 1;
    else return mid;
  }
  return -1;
}

export interface DecodedHeading {
  readonly id: string;
  readonly text: string;
  readonly level: number;
  readonly section: number;
}

export function decodeHeadings(
  buffer: OpBuffer,
  strings: OpStrings,
): DecodedHeading[] {
  const text = strings.strings;
  const headings: DecodedHeading[] = [];
  for (let i = 0; i < buffer.headings.length; i += HEADING_WORDS)
    headings.push({
      level: buffer.headings[i],
      id: text.substring(
        buffer.headings[i + 1],
        buffer.headings[i + 1] + buffer.headings[i + 2],
      ),
      section: buffer.headings[i + 3],
      text: text.substring(
        buffer.headings[i + 4],
        buffer.headings[i + 4] + buffer.headings[i + 5],
      ),
    });
  return headings;
}

/**
 * Where the attribute and element policy of the replay lives. The buffer
 * guarantees the tag and attribute *names*; the values stay the sink's
 * responsibility, exactly as in `sanitizeFragment`.
 */
export interface ReplaySink {
  /** Called in document order, before `opened` for the same element. */
  attribute(element: Element, name: string, value: string): void;
  /** Called once the element carries its attributes. */
  opened(element: Element): void;
  /**
   * Called for every text node with its offset in the document text. This is
   * the mapping the search uses to turn a text offset back into a DOM position
   * without ever indexing the DOM.
   */
  text?(node: Text, offset: number, length: number): void;
}

/** Structural replay without any value policy; for round-trip tests. */
export const structuralSink: ReplaySink = {
  attribute: (element, name, value) => element.setAttribute(name, value),
  opened: () => {},
};

export function replaySection(
  buffer: OpBuffer,
  strings: OpStrings,
  index: number,
  sink: ReplaySink = structuralSink,
  owner: Document = document,
): DocumentFragment {
  const fragment = owner.createDocumentFragment();
  const base = index * SECTION_WORDS;
  const start = buffer.sections[base];
  const end = start + buffer.sections[base + 1];
  const { ops, attrs } = buffer;
  const stack: (Element | DocumentFragment)[] = [fragment];
  let depth = 0;
  for (let i = start; i < end; i++) {
    const op = i * OP_WORDS;
    switch (ops[op]) {
      case OP_OPEN: {
        const element = owner.createElement(ALLOWED_TAGS[ops[op + 1]]);
        const attrEnd = ops[op + 2] + ops[op + 3];
        for (let a = ops[op + 2]; a < attrEnd; a++) {
          const word = a * ATTR_WORDS;
          sink.attribute(
            element,
            ALLOWED_ATTR[attrs[word]],
            strings.strings.substring(
              attrs[word + 1],
              attrs[word + 1] + attrs[word + 2],
            ),
          );
        }
        sink.opened(element);
        stack[depth].append(element);
        stack[++depth] = element;
        break;
      }
      case OP_CLOSE:
        depth--;
        break;
      default: {
        const node = owner.createTextNode(
          strings.text.substring(ops[op + 1], ops[op + 1] + ops[op + 2]),
        );
        stack[depth].append(node);
        sink.text?.(node, ops[op + 1], ops[op + 2]);
      }
    }
  }
  return fragment;
}

File: src-tauri/markdown/src/opbuffer.rs
//! The op buffer from docs/decisions/007-performance-path.md, section 4.
//!
//! The layout is the one `src/core/markdown/opbuffer.ts` documents and decodes.
//! Offsets and lengths are **UTF-16 code units** into the string the renderer
//! obtains by decoding `strings` once per document — never byte offsets. The
//! encoder carries that count forward while it appends, which is the whole
//! reason the renderer can use `substring` without a translation table.

pub const OP_OPEN: u32 = 0;
pub const OP_CLOSE: u32 = 1;
pub const OP_TEXT: u32 = 2;
pub const SECTION_FALLBACK: u32 = 1;

/// Exactly the tag allowlist `opbuffer.ts` exports; a tag outside it has no id
/// and is therefore not expressible in the buffer at all.
pub const ALLOWED_TAGS: [&str; 39] = [
    "p", "h1", "h2", "h3", "h4", "h5", "h6", "em", "strong", "del", "s", "code", "pre",
    "blockquote", "ul", "ol", "li", "hr", "br", "a", "img", "table", "thead", "tbody", "tfoot",
    "tr", "th", "td", "input", "sup", "sub", "kbd", "details", "summary", "div", "span", "dl", "dt",
    "dd",
];
/// Exactly the attribute allowlist `opbuffer.ts` exports.
pub const ALLOWED_ATTR: [&str; 16] = [
    "id", "href", "src", "alt", "title", "class", "start", "type", "checked", "disabled", "colspan",
    "rowspan", "align", "width", "height", "open",
];

pub const TAG_P: u32 = 0;
pub const TAG_EM: u32 = 7;
pub const TAG_STRONG: u32 = 8;
pub const TAG_DEL: u32 = 9;
pub const TAG_CODE: u32 = 11;
pub const TAG_PRE: u32 = 12;
pub const TAG_BLOCKQUOTE: u32 = 13;
pub const TAG_UL: u32 = 14;
pub const TAG_OL: u32 = 15;
pub const TAG_LI: u32 = 16;
pub const TAG_HR: u32 = 17;
pub const TAG_BR: u32 = 18;
pub const TAG_A: u32 = 19;
pub const TAG_IMG: u32 = 20;
pub const TAG_TABLE: u32 = 21;
pub const TAG_THEAD: u32 = 22;
pub const TAG_TBODY: u32 = 23;
pub const TAG_TR: u32 = 25;
pub const TAG_SUMMARY: u32 = 33;
pub const TAG_DT: u32 = 37;
pub const TAG_DD: u32 = 38;
pub const TAG_TH: u32 = 26;
pub const TAG_TD: u32 = 27;
pub const TAG_INPUT: u32 = 28;
pub const TAG_SUP: u32 = 29;
pub const TAG_DIV: u32 = 34;

pub const ATTR_ID: u32 = 0;
pub const ATTR_HREF: u32 = 1;
pub const ATTR_SRC: u32 = 2;
pub const ATTR_ALT: u32 = 3;
pub const ATTR_TITLE: u32 = 4;
pub const ATTR_CLASS: u32 = 5;
pub const ATTR_START: u32 = 6;
pub const ATTR_TYPE: u32 = 7;
pub const ATTR_CHECKED: u32 = 8;
pub const ATTR_DISABLED: u32 = 9;
pub const ATTR_ALIGN: u32 = 12;

fn utf16_len(value: &str) -> u32 {
    if value.is_ascii() {
        value.len() as u32
    } else {
        value.chars().map(|c| c.len_utf16() as u32).sum()
    }
}

/// FNV-1a. Used only for section identity (see 008, section 5), never for
/// integrity or security.
#[derive(Clone, Copy)]
pub struct Fnv(u64);
impl Default for Fnv {
    fn default() -> Self {
        Fnv(0xcbf2_9ce4_8422_2325)
    }
}
impl Fnv {
    pub fn write(&mut self, bytes: &[u8]) {
        for &byte in bytes {
            self.0 ^= byte as u64;
            self.0 = self.0.wrapping_mul(0x1000_0000_01b3);
        }
    }
    pub fn write_u32(&mut self, value: u32) {
        self.write(&value.to_le_bytes());
    }
    pub fn value(self) -> u64 {
        self.0
    }
}

// The blocks `indexText` treated as separate for search, unchanged: a match
// must not run from the end of one block into the start of the next.
const BLOCK_TAGS: [u32; 14] = [
    TAG_P, TAG_LI, TAG_PRE, TAG_TD, TAG_TH, 1, 2, 3, 4, 5, 6, TAG_SUMMARY, TAG_DT, TAG_DD,
];

#[derive(Default)]
pub struct Encoder {
    pub ops: Vec<u32>,
    pub attrs: Vec<u32>,
    /// Attribute values, heading ids and texts, and the HTML of fallback
    /// sections. Never document text.
    pub strings: String,
    units: u32,
    /// The document's text in document order — what the search runs on.
    pub text: String,
    text_units: u32,
    pub attr_count: u32,
    hash: Fnv,
    pending: String,
    /// Serial of the block element enclosing the current position, and of the
    /// one that produced the previous text run.
    blocks: Vec<u32>,
    block_serial: u32,
    last_block: Option<u32>,
}

impl Encoder {
    fn intern(&mut self, value: &str) -> u32 {
        let offset = self.units;
        self.strings.push_str(value);
        self.units += utf16_len(value);
        offset
    }
    fn current_block(&self) -> Option<u32> {
        self.blocks.last().copied()
    }
    /// Length of the document text so far, in UTF-16 code units.
    pub fn text_units(&self) -> u32 {
        self.text_units
    }
    /// Separates two sections whose text must not run together, for sections
    /// that contribute no operations at all.
    pub fn separate(&mut self) {
        self.text.push('\n');
        self.text_units += 1;
        self.last_block = None;
    }
    /// Text runs are merged the way an HTML parser merges them, so that
    /// `isEqualNode` stays usable as a contract and fewer nodes are created.
    pub fn text(&mut self, value: &str) {
        self.pending.push_str(value);
    }
    fn flush_text(&mut self) {
        if self.pending.is_empty() {
            return;
        }
        let pending = std::mem::take(&mut self.pending);
        self.hash.write(b"\x02");
        self.hash.write(pending.as_bytes());
        let block = self.current_block();
        if self.last_block.is_some() && block != self.last_block {
            // The separator belongs to no operation; it only keeps a match from
            // spanning a block boundary, exactly as `indexText` did.
            self.text.push('\n');
            self.text_units += 1;
        }
        self.last_block = block;
        let offset = self.text_units;
        self.text.push_str(&pending);
        let length = utf16_len(&pending);
        self.text_units += length;
        self.ops.extend_from_slice(&[OP_TEXT, offset, length, 0]);
        self.pending = pending;
        self.pending.clear();
    }
    pub fn open(&mut self, tag: u32) -> Open {
        self.flush_text();
        self.hash.write(b"\x00");
        self.hash.write_u32(tag);
        // Every open pushes the block in force inside it, so the enclosing
        // block is one lookup rather than a walk up the stack.
        let block = if BLOCK_TAGS.contains(&tag) {
            self.block_serial += 1;
            Some(self.block_serial)
        } else {
            self.current_block()
        };
        self.blocks.push(block.unwrap_or(0));
        Open {
            tag,
            attr_start: self.attr_count,
        }
    }
    pub fn attribute(&mut self, open: &Open, name: u32, value: &str) {
        debug_assert!(open.attr_start <= self.attr_count);
        self.hash.write(b"\x03");
        self.hash.write_u32(name);
        self.hash.write(value.as_bytes());
        let offset = self.intern(value);
        let length = utf16_len(value);
        self.attrs.extend_from_slice(&[name, offset, length]);
        self.attr_count += 1;
    }
    /// Emits the OPEN once its attributes are known.
    pub fn opened(&mut self, open: Open) {
        let count = self.attr_count - open.attr_start;
        self.ops
            .extend_from_slice(&[OP_OPEN, open.tag, open.attr_start, count]);
    }
    pub fn close(&mut self) {
        self.flush_text();
        self.hash.write(b"\x01");
        self.blocks.pop();
        self.ops.extend_from_slice(&[OP_CLOSE, 0, 0, 0]);
    }
    pub fn end_section(&mut self) -> (usize, u64) {
        self.flush_text();
        let hash = self.hash.value();
        self.hash = Fnv::default();
        (self.ops.len() / 4, hash)
    }
    /// Offset and length of `value` after appending it; for ids and heading text.
    pub fn reference(&mut self, value: &str) -> (u32, u32) {
        (self.intern(value), utf16_len(value))
    }
    pub fn hash_bytes(&mut self, bytes: &[u8]) {
        self.hash.write(bytes);
    }
    pub fn take_hash(&mut self) -> u64 {
        let hash = self.hash.value();
        self.hash = Fnv::default();
        hash
    }
}

/// An element whose attributes are still being collected.
pub struct Open {
    tag: u32,
    attr_start: u32,
}
